ServerCore/Pages/Shared/_RoleSwitcherPartial.cshtml

```diff
@@ -1,24 +1,40 @@
-    var IsRegistered = (Model as ServerCore.ModelBases.EventSpecificPageModel)?.IsRegisteredUser().Result ?? false;
-    var AdminRoute = ViewData["AdminRoute"] as string ?? "/EventSpecific/Index";
-    var AuthorRoute = ViewData["AuthorRoute"] as string ?? "/EventSpecific/Index";
-    var PlayRoute = ViewData["PlayRoute"] as string ?? "/EventSpecific/Index";
-    if (!IsRegistered && !PlayRoute.StartsWith("/EventSpecific/") && !PlayRoute.Equals("/Teams/Apply") && !PlayRoute.Equals("/Teams/Create"))
+    var EventModel = Model as ServerCore.ModelBases.EventSpecificPageModel;
+
+    if (EventModel != null)
     {
-        // If the user isn't registered and isn't on an event page or the two teams pages they have access to, always redirect to the registration page
-        PlayRoute = "/Teams/List";
-    }
-    var RoutingPuzzleId = ViewData["RoutingPuzzleId"] as string;
-    var RoutingTeamId = ViewData["RoutingTeamId"] as string;
-}
+        var Event = EventModel.Event;
+        var EventRole = EventModel.EventRole;
+        var IsAdmin = EventRole == ModelBases.EventRole.admin || await EventModel.IsEventAdmin();
+        var IsAuthor = EventRole == ModelBases.EventRole.author || await EventModel.IsEventAuthor();
+
+        if (IsAdmin || IsAuthor)
+        {
+            var IsRegistered = await EventModel.IsRegisteredUser();
+            var AdminRoute = ViewData["AdminRoute"] as string ?? "/EventSpecific/Index";
+            var AuthorRoute = ViewData["AuthorRoute"] as string ?? "/EventSpecific/Index";
+            var PlayRoute = ViewData["PlayRoute"] as string ?? "/EventSpecific/Index";
+            if (!IsRegistered && !PlayRoute.StartsWith("/EventSpecific/") && !PlayRoute.Equals("/Teams/Apply") && !PlayRoute.Equals("/Teams/Create"))
+            {
+                // If the user isn't registered and isn't on an event page or the two teams pages they have access to, always redirect to the registration page
+                PlayRoute = "/Teams/List";
+            }
+            var RoutingPuzzleId = ViewData["RoutingPuzzleId"] as string;
+            var RoutingTeamId = ViewData["RoutingTeamId"] as string;
 
-<!-- TODO: Check for permissions -->
 <ul class="dropdown-menu">
+    @if (IsAdmin)
+    {
     <li><a asp-page=@AdminRoute asp-route-id="@Event.ID" asp-route-eventRole="@ModelBases.EventRole.admin" asp-route-teamId=@RoutingTeamId asp-route-puzzleId=@RoutingPuzzleId>Admin mode</a></li>
     <li class="divider"></li>
+    }
+    @if (IsAuthor)
+    {
     <li><a asp-page=@AuthorRoute asp-route-id="@Event.ID" asp-route-eventRole="@ModelBases.EventRole.author" asp-route-teamId=@RoutingTeamId asp-route-puzzleId=@RoutingPuzzleId>Author mode</a></li>
     <li class="divider"></li>
+    }
     <li><a asp-page=@PlayRoute asp-route-id="@Event.ID" asp-route-eventRole="@ModelBases.EventRole.play" asp-route-teamId=@RoutingTeamId asp-route-puzzleId=@RoutingPuzzleId>Player mode</a></li>
```

Comment:
We'll want to hide this item / possibly hide the dropdown overall if the user isn't an author or an admin - users who are just players don't need to know that role switching exists and a non-functional dropdown could be confusing. We can tackle that separately though if you'd like - this is still a huge step forward :)

Reply:
Note the IsAdmin || IsAuthor check at line 11 - we do hide the dropdown in those cases.

Reply:
See line 11 above - I already added this check.